code/config/lib/Loot/Config/Generics.hs

```diff
@@ -0,0 +1,176 @@
+{- SPDX-License-Identifier: MPL-2.0 -}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+{-# LANGUAGE ConstraintKinds      #-}
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE KindSignatures       #-}
+{-# LANGUAGE MonoLocalBinds       #-}
+{-# LANGUAGE PolyKinds            #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- | Utilities for converting between configuration and ADT.
+
+Example: let's assume you have the following datatype:
+
+@
+data DBConfig = DBConfig
+  { username :: Text
+  , password :: Text
+  , port :: Word16
+  } deriving Generic
+@
+
+Then it can be included into configuration as follows:
+
+@
+type Config =
+  '[ "myservice" ::<
+      '[ "db" ::< 'FromData' DBConfig
+       ]
+   -- This unfolds into
+   -- '[ "db" ::<
+   --     [ "username" ::: Text
+   --     , "password" ::: Text
+   --     , "port" ::: Word16
+   --     ]
+   --  ]
+   ]
+
+dbConfig :: ConfigRec 'Final Config -> DBConfig
+dbConfig cfg = cfg ^. sub #myservice . sub #db . 'asData'
+@
+
+-}
+module Loot.Config.Generics
+       ( FromData
+       , toData
+       , fromData
+       , asData
+
+         -- * Internal helpers
+       , KnownConfigKind
+       ) where
+
+import Data.Vinyl (Rec ((:&), RNil), (<+>))
+import Data.Vinyl.TypeLevel (type (++))
+import GHC.TypeLits (ErrorMessage (..), TypeError)
+
+import Loot.Config.Record ((:::), ConfigKind (Final, Partial), ConfigRec,
```

Comment:
Shouldn't it be the last section of imports?

Reply:
Right, fixed.